tsmf updates/fixes for gstreamer decoder on X11 platform

bjcollins opened the pull request:
- Use decodebin2 instead of old decodebin - decodebin has issues
- Use autovideosink - xvimagesink does not work with cards with no xv ports available and cant be used if wanted to use the fluendo hardware accelerated playback codec
- Use autoaudiosink - let gstreamer choose best audio playback plugin
- Catch when autosinks add known elements so that we can manipulate properties on them
- Adjust caps of various media types to work better with gstreamer, some codecs are picky about having certain fields available
- Remove unneeded plugins such as "ffmpegcolorspace" and "videoscale" - these do not work correctly with fluendo hardware accelerated playback codec
- Name audio/video gstreamer elements better for easier debugging
- Update gstreamer pipeline and element properties to handle playback better
- Detect when valid timestamps are available for buffer from server and try to account for when they are not valid
- Start time is much more reliable then end time from server for various media formats, so use it when possible to make decisions instead of end time
- Do not rebuild gstreamer pipeline for a seek(very expensive), instead reset gstreamer time to 0 and maintain offset between real time and gstreamer time
- Change buffer filled function back to a buffer level function, so that we can use buffer level to make better choices above gstreamer decoder in tsmf
- Remove ack function from gstreamer, instead rely on ack thread to handle acks
- Rework X11 gstreamer code to handle various videosinks which implement the XOverlayInterface and to keep more detailed information on the sub-window that is used for display
- Add check to see if a decoder is available for telling the server the client various media types
- Add in support for M4S2 and WMA1 media types
- Fix flush message handling, they are for individual streams and not the entire presentation
- Delay eos response to try to allow more time for buffers to be loaded into decoder, as we anticipate acks to server and the server will issue stop as soon as we ack eos.
- Fix issue with geometry info being ignored when resent for new streams within existing presentation
- Fixed volume level initialization issue when a stream is stopped and restarted
- Attempt to sync video/audio streams...because we run two different gstreamer pipelines - they can enter pause/playing states at different times and are thus not synchronized. Attempt
to adjust video buffer timestamps based on difference between audio/video running time to account for this difference. This logic accounts for a huge improvement in audio/video sync(ie. lip sync to words)

hardening commented on channels/tsmf/client/gstreamer/tsmf_X11.c:
```diff
@@ -69,13 +75,73 @@ static const char* get_shm_id()
 const char* tsmf_platform_get_audio_sink(void)
 {
+	//return "alsasink";
 	return "autoaudiosink";
```
comment to drop

bjcollins replied:
agreed

hardening commented on channels/tsmf/client/gstreamer/tsmf_gstreamer.c:
```diff
@@ -43,10 +43,14 @@
 #endif
 
+// 1 second
+#define SEEK_TOLERANCE 10000000
```
no C++ comment
1 * 1000 * 1000 would be more readable ?

bjcollins replied:
Oh, I think I have other comments that need to be multiline as well, I can do your suggestion to make the value more readable.

hardening commented on channels/tsmf/client/gstreamer/tsmf_gstreamer.c:
```diff
@@ -55,8 +59,33 @@ const char* get_type(TSMFGstreamerDecoder* mdecoder)
-	else
+	else if (mdecoder->media_type == TSMF_MAJOR_TYPE_AUDIO)
 		return "AUDIO";
+	else
```
else can be dropped

bjcollins replied:
I think we should keep the else, UNKNOWN is a valid value and it is possible for a decoder object to have that value for the media_type parameter if it has not been initialized.

hardening replied:
Well in fact this should be a switch statement !

bjcollins replied:
Ok, no problem. I will change it to a switch statement.

hardening commented on channels/tsmf/client/gstreamer/tsmf_gstreamer.c:
```diff
@@ -750,7 +942,7 @@ BOOL tsmf_gstreamer_ack(ITSMFDecoder* decoder, BOOL (*cb)(void *, BOOL), void *s
 	DEBUG_TSMF("");
-	mdecoder->ack_cb = cb;
+	mdecoder->ack_cb = NULL;//cb;
 	mdecoder->stream = stream;
```
drop comment

bjcollins replied:
agreed, should probably just remove the whole ack_cb stuff from gstreamer decoder code.

hardening commented on channels/tsmf/client/tsmf_main.c:
```diff
@@ -34,6 +34,38 @@
+
+	if (callback->stream_id)
+	{
+		s = Stream_New(NULL, 24);
```
return value not checked

bjcollins replied:
Ok, just for reference - there are other areas in tsmf that do not check the return value for this call either. I will try to hit all cases with an updated pull request.

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -238,6 +274,8 @@ static void tsmf_sample_queue_ack(TSMF_SAMPLE* sample)
 
+// Returns TRUE if no more samples are currently available
+// Returns FALSE otherwise
 static BOOL tsmf_stream_process_ack(void* arg, BOOL force)
```
no C++ comment

bjcollins replied:
Agreed

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -246,22 +284,38 @@ static BOOL tsmf_stream_process_ack(void* arg, BOOL force)
 	if (!stream)
-		return FALSE;
+		return TRUE;
 
```
are you sure about this one ?

bjcollins replied:
I think this is correct, it matches my description for the return code. I use TRUE to indicate that no samples are currently available, which would be the case if there was no stream. Make sense at all?

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -882,7 +1026,7 @@ void tsmf_presentation_set_audio_device(TSMF_PRESENTATION* presentation, const c
 }
 
-static void tsmf_stream_flush(TSMF_STREAM* stream)
+void tsmf_stream_flush(TSMF_STREAM* stream)
```
is the static removed on purpose ?

bjcollins replied:
Yes, I am now calling this function from tsmf_ifman.c t support flush message per stream instead of for the entire presentation.

hardening commented:
@freerdp-bot would  you be so kind and test this nice PR ?

hardening commented:
@bjcollins nice PR.

Guys with the merge rights, we have a branch that contains tons of fixes for the TSMF channels, some are possibly conflicting with this PR. I'd suggest that we wait before merging this one.

bjcollins commented:
I looked at the output and I think my PR is probably flawed in terms of gstreamer 1.0 support as I was focused on gstreamer 0.10, I will look at fixing and resubmitting along with addressing various comments made against the PR.

hardening commented:
@bjcollins don't invest too much in checks, we have a PR addressing all the checks that should be done and aren't

bjcollins commented:
Ok. I just did something quick checks based on those few Stream_New calls, finished incorporating your comments also. But need to make fixes and do some Gstreamer 1.0 testing. I made some changes that were not compatible across both versions of gstreamer. Hoping to resubmit today.

bjcollins commented:
Ok. I made some changes based on the feedback so far and fixed the gstreamer 1.0 build issue from my first commit. Retested both a gstreamer 0.10 and gstreamer 1.0 build.

akallabeth commented:
lgtm, +1

bjcollins commented:
Do I need to do something more to resolve the conflict? My original checkin failed, but I have since commited a fix to that.

akallabeth commented:
@bjcollins Looks like you need to rebase your branch.

bjcollins commented:
Ok, I updated I fixed this properly now.

@akallabeth Thanks for hint.  I do not use git much, but I also did not realize I was actually in conflict. Just thought it was still saying it was broke due to my first broke commit.

bjcollins commented:
Rebased this again. Hopefully, it can be reconsidered now.

akallabeth commented:
@awakecoding @hardening @bmiklautz This branch looks good to me, is there something preventing merging?

hardening commented:
Nope, It looks good to me. We were waiting for the channel hardening to be in to consider merging this one. Now that it's upstream, let's merge.
